File: pkg/forky/api/ethereum.go
package api

import (
	"context"
	"encoding/json"
	"net/http"

	fhttp "github.com/ethpandaops/forky/pkg/forky/api/http"

	"github.com/julienschmidt/httprouter"
)

func (h *HTTP) handleV1GetEthereumSpec(ctx context.Context, _ *http.Request, _ httprouter.Params, contentType fhttp.ContentType) (*fhttp.Response, error) {
	if err := fhttp.ValidateContentType(contentType, []fhttp.ContentType{fhttp.ContentTypeJSON}); err != nil {
		return fhttp.NewUnsupportedMediaTypeResponse(nil), err
	}

	rsp := fhttp.V1GetEthereumSpecResponse{
		NetworkName: h.svc.GetEthereumNetworkName(ctx),
		Spec: fhttp.EthereumSpec{
			SecondsPerSlot: h.svc.GetEthereumSpecSecondsPerSlot(ctx),
			SlotsPerEpoch:  h.svc.GetEthereumSpecSlotsPerEpoch(ctx),
			GenesisTime:    h.svc.GetEthereumSpecGenesisTime(ctx),
		},
	}

	response := fhttp.NewSuccessResponse(fhttp.ContentTypeResolvers{
		fhttp.ContentTypeJSON: func() ([]byte, error) {
			return json.Marshal(rsp)
		},
	})

	response.SetCacheControl("public, max-age=60, s-maxage=60")

	return response, nil
}

func (h *HTTP) handleV1GetEthereumNow(ctx context.Context, _ *http.Request, _ httprouter.Params, contentType fhttp.ContentType) (*fhttp.Response, error) {
	if err := fhttp.ValidateContentType(contentType, []fhttp.ContentType{fhttp.ContentTypeJSON}); err != nil {
		return fhttp.NewUnsupportedMediaTypeResponse(nil), err
	}

	slot, epoch, err := h.svc.GetEthereumNow(ctx)
	if err != nil {
		return fhttp.NewInternalServerErrorResponse(nil), err
	}

	rsp := fhttp.V1GetEthereumNowResponse{
		Slot:  uint64(slot),
		Epoch: uint64(epoch),
	}

	response := fhttp.NewSuccessResponse(fhttp.ContentTypeResolvers{
		fhttp.ContentTypeJSON: func() ([]byte, error) {
			return json.Marshal(rsp)
		},
	})

	response.SetCacheControl("public, max-age=1, s-maxage=1")

	return response, nil
}

File: web/src/api/ethereum.ts
import {
  Response,
  V1GetEthereumNowResponse,
  V1GetEthereumSpecResponse,
  EthereumSpec,
} from '@app/types/api';
import { BASE_URL } from '@utils/environment';

export type Spec = {
  network_name: string;
  spec: EthereumSpec;
};

export async function fetchNow(): Promise<Required<V1GetEthereumNowResponse>> {
  const response = await fetch(`${BASE_URL}api/v1/ethereum/now`);

  if (!response.ok) {
    throw new Error('Failed to fetch slot data');
  }
  const json = (await response.json()) as Response<V1GetEthereumNowResponse>;

  if (!json.data?.slot) throw new Error('No slot in response');
  if (!json.data?.epoch) throw new Error('No epoch in response');

  return {
    slot: json.data?.slot,
    epoch: json.data?.epoch,
  };
}

export async function fetchSpec(): Promise<Required<Spec>> {
  const response = await fetch(`${BASE_URL}api/v1/ethereum/spec`);

  if (!response.ok) {
    throw new Error('Failed to fetch slot data');
  }
  const json = (await response.json()) as Response<V1GetEthereumSpecResponse>;

  if (!json.data?.spec) throw new Error('No spec in response');

  return {
    network_name: json.data?.network_name ?? 'unknown',
    spec: json.data.spec,
  };
}
